feat: SMTP Templates

doncicuto opened the pull request:
With this feature you are able to choose an SMTP template and default settings are already made, so you only have to give my configuration specific settings.

In this PR there are default settings for Amazon SES, Mailgun, Mailjet, Postmark, Sendgrid and a Generic SMTP provider. If you have known settings for other providers let me known and I'll add them during this review.

This is a video showing this feature:

https://github.com/zitadel/zitadel/assets/30386061/a32b9f5d-a666-46e2-9cdc-efabbfa24a5e

This PR has many frontend/backend changes, some of them are breaking changes:

- As we now are able to have many SMTP configurations an id must be used to specify the config to be updated, removed, activated or deactivated or if you want to change SMTP password. Also we must activate a provider in order to send emails.
- golangci-lint contextcheck complained about passing context to InstanceAggregateFromWriteModel.
- I've kept UpdateSMTPConfigPassword in API but when you create or edit an SMTP config you can send a password.
- GetSMTPConfig API will give us the current ACTIVE configuration.

I've done my best, but I guess that many things will be wrong in backend as I'm still trying to learn more about it. I thank you in advance for your review

Closes #5503

### Definition of Ready

- [X] I am happy with the code
- [X] Short description of the feature/issue is added in the pr description
- [X] PR is linked to the corresponding user story
- [X] Acceptance criteria are met
- [ ] All open todos and follow ups are defined in a new ticket and justified
- [ ] Deviations from the acceptance criteria and design are agreed with the PO and documented.
- [X] No debug or dead code
- [ ] My code has no repetitions
- [X] Critical parts are tested automatically
- [X] Where possible E2E tests are implemented
- [X] Documentation/examples are up-to-date
- [ ] All non-functional requirements are met
- [X] Functionality of the acceptance criteria is checked manually on the dev system.

muhlemmer commented:
Thanks for just another great PR! @peintnermax  is quite busy this week with some internal stuff and I imagine he would have more time later this week or next week.

hifabienne commented:
hei @doncicuto 
Thanks for the pr, we agreed internally that i will run it locally to verify from a business perspective that it does what it should, before we start reviewing from a technical perspective.
I am pretty busy this week, but will do as soon as possible.

doncicuto commented:
Sure @hifabienne! Hope It works as you expected It to be and its useful overall 😄 Thanks!

eliobischof reviewed (changes requested):
Again great job @doncicuto :pray: 
I'm sure having SMTP templates greatly improves the UX.

I merged back main and tested it. On a fresh instance, everything works fine.
However, I created an SMTP config using the current code on main and then ran the code of this PR. It looks like something is broken with the `provider_type` in the projection.

![image](https://github.com/zitadel/zitadel/assets/12727842/39a07390-8ad7-462e-a28b-b7f0576f895a)
![image](https://github.com/zitadel/zitadel/assets/12727842/1626643c-40a2-4798-98c6-71cf44e1f930)
I think the database column type should be `handler.ColumnTypeInt64` instead of `handler.ColumnTypeEnum`. However, IMO switching the generic provider to have the zero value `0` as index would make sense to me.

I tend to like the decision of keeping the provider type as just a loose integer in the ZITADEL API, as the notification channel needs the same config structure for any type, at least ATM. Like this, no domain foreign logic is leaked into the API. This can easily be added in the future if needed. Anyway, I'd like to sleep on it and come back to you about this.

The translation changes look a bit wild :smiley: Did you add some missing translations on the go or is that unintended?
BTW, in the meantime, dutch translations, i18n nl files, were added to main and missing in the PR.

Looking forward to merge this :heart: :rocket:

doncicuto commented:
Hi @eliobischof, first of all thanks for the review and your patience, I know this is a huge PR and it'll take some time before it can be merged. Hopefully @hifabienne found this approach useful to the project.

About translations, sure I'll add the dutch translation. And, indeed, I added some missing translations for french, chinese and italian after I found that these files have a lot of missing lines (maybe I should have done that in a different PR, sorry about that).

About column types, I'm just a noob with the backend and have no strong opinion so whatever you feel best based on your experience it'll work for me (I really appreciate this kind of explanations for future PR). In any case your initial thoughts sound sensible to me as I also want that old Zitadel versions configs get the generic template association. Just let me know your final decision on this and I'll change the code accordingly.

eliobischof reviewed (changes requested):
@doncicuto, here is what @stebenz and I discussed:

- Let's remove the provider type field in the backend, as the backend shouldn't care to which provider the config belongs. This means, in the configured providers list in the console, we can remove the icons and leave it to the user to configure a meaningful provider name.
- We can also get rid of the new ID field of the provider and use the aggregate ID instead.

Does this make sense, so far? Please shout if you need help.

doncicuto commented:
Hi @eliobischof and @stebenz thanks for your comments about the provider type and the provider id. I'll start preparing them. If I have any doubt I'll let you know.

Thanks!

doncicuto commented:
Hi @eliobischof and @stebenz. I'm removing the provider_type field as requested and adding a description field that admins can use to give a descriptive name for provider. 

I've one doubt about removing the provider ID. We want to have several SMTP providers for an instance (only one active at a time, but the rest are there if you want to activate it later). Can the aggregate ID be used to uniquely identify each provider that I add? I've based this code on the existing SMS provider code that is ready to have several providers for every instance and this code use an ID: https://github.com/zitadel/zitadel/blob/main/internal/repository/instance/sms.go#L22C1-L29C2

Initialy I tried to use that aggregate ID field, but I remember that it didn't help to have different providers so I finally opted to add the ID field.

Thanks!

doncicuto commented:
Hi there! I know you're busy so whenever you've time I need to discuss some doubts.

stebenz commented:
> Hi there! I know you're busy so whenever you read this and have some spare time I need to discuss some doubts.

Hi @doncicuto, sorry for the late reply, whenever you have the time we can arrange a meeting to discuss your doubts, if still necessary.

To your previously asked question, if you look into the file https://github.com/zitadel/zitadel/blob/main/internal/command/sms_config.go#L32 you will see that the ID is to identify the current configuration and the aggregate ID is the instance as it is a 1-1 connection.
That's exactly why we would use a separate aggregate ID for the providers, as we would have multiple per instance and you would need no ID in the event in that case.

livio-a commented on internal/domain/smtp.go:
```diff
@@ -5,5 +5,10 @@ type SMTPConfigState int32
 const (
 	SMTPConfigStateUnspecified SMTPConfigState = iota
 	SMTPConfigStateActive
+	SMTPConfigStateInactive
```
just noted by coincidence: this change would effect existing smtp configuration, which were removed. they would be considered inactive instead

doncicuto replied:
Thanks @livio-a, I've moved the new inactive state after existing removed state. Whenever I play with iota in the future I'll be more careful :sweat_smile:

hifabienne commented:
@doncicuto do you still need help from our side?

doncicuto commented:
Hey @hifabienne first of all I'm sorry that this PR is not ready yet. 

I'll try to review its status on Sunday. First I'll try to fix the conflicts that blocks the merge. Second I'll add the code that uses only aggregation_id as the SMTP settings id. Then I'll need the help of the team because I tried to test my code against an old zitadel container with SMTP settings and I have some problems as those old settings are not migrated with the right info.

I love to try backend issues but I don't fully understand the eventstore, migrations... I wish to be more productiva, do you have any good docs/links about this pattern that goes beyond   the info you have in the docs?

I have some issues right now at my job so I only have time to work on Zitadel the weekend or with easy issues during the week. I Hope this weekend I can prepare this PR so you can help me to finish it. Thanks!

doncicuto commented:
Hi @hifabienne. I need some help here to finish this.

We want to have several SMTP settings available for an instance. Only one setting will be active.

Here's my proposal for the new projection:

```
const (
	SMTPConfigProjectionTable = "projections.smtp_configs2"
        SMTPConfigColumnID = "id"
	SMTPConfigColumnAggregateID    = "aggregate_id"
	SMTPConfigColumnCreationDate   = "creation_date"
	SMTPConfigColumnChangeDate     = "change_date"
	SMTPConfigColumnSequence       = "sequence"
	SMTPConfigColumnDescription    = "description"
	SMTPConfigColumnResourceOwner  = "resource_owner"
	SMTPConfigColumnInstanceID     = "instance_id"
	SMTPConfigColumnTLS            = "tls"
	SMTPConfigColumnSenderAddress  = "sender_address"
	SMTPConfigColumnSenderName     = "sender_name"
	SMTPConfigColumnReplyToAddress = "reply_to_address"
	SMTPConfigColumnSMTPHost       = "host"
	SMTPConfigColumnSMTPUser       = "username"
	SMTPConfigColumnSMTPPassword   = "password"
	SMTPConfigColumnState          = "state"
)
```

Please, could you confirm me that you're ok if I have a SMTPConfigColumnID that identifies each setting? or instead you want to use only SMTPConfigColumnAggregateID and SMTPConfigColumnInstanceID as a primary key. In a future PR we'd like to have smtp settings per org so maybe this id would be fine.

Once you confirm me that I'll continue with my questions

Thanks in advance

stebenz commented:
@doncicuto maybe I can help here.
Unfortunately, some comparable topics came up internally a lot at the moment, I can't really tell you a definite answer but as a starting point, we discussed the idea of moving the SMTP config in a separate aggregate, which would eliminate this question about the additional ID but would mean a lot of other changes.

I will try to pick this topic up next week with @adlerhurst and @muhlemmer on how to handle such changes and will get back to you so that we can have maybe a look at it together.

doncicuto commented:
Thanks @stebenz this topic is difficult to me as I have little knowledge about the eventstore so I appreciate your time and patience. Any hints you can give me once you discuss it will help me to finish this PR.

stebenz commented:
@doncicuto Good news, we decided to go with the faster option, please use a combination of InstanceID+ResourceOwner+ID as primary key for the projection, so it would also be possible to add provider on org level later, which means you can remove aggregate_id.
The ID will not have a default value as you have in your PR, but that can be filled also from the "old" events without ID with the resourceOwner.
I hope this makes it a little clearer, would also be happy to hop on a call if necessary.
If you have other questions or need somebody to do a review just tell us.

doncicuto commented:
Thanks a lot @stebenz I'll prepare the code following these guidelines. I appreciate the chance of having that call if I find further problems but I hope that I'll be able to make It work.

doncicuto commented:
Ok @hifabienne I think this is ready for review. I've done my best but I think it's time for senior staff to check what can be done better. I don't think I can make further progress until the team can give me more code samples as I haven't a deep knowledge of backend yet. Hope the team don't sweat too much reviewing this PR. Thanks in advance

stebenz commented on internal/query/projection/smtp.go:
```diff
@@ -11,8 +12,8 @@ import (
-
+	SMTPConfigProjectionTable      = "projections.smtp_configs2"
+	SMTPConfigColumnID             = "id"
 	SMTPConfigColumnAggregateID    = "aggregate_id"
```
So aggregateID column is not used anymore?

doncicuto commented on internal/query/smtp.go:
```diff
@@ -73,40 +84,71 @@ var (
+		return nil, zerrors.ThrowInternal(err, "QUERY-3m9sl", "Errors.Query.SQLStatement")
+	}
+
+	err = q.client.QueryRowContext(ctx, func(row *sql.Row) error {
```
Hi @stebenz. In current Zitadel code SMTPConfigByAggregateID is used to get the only and unique smtp config for the instance that is going to be used to send emails. In this PR, an instance can have multiple smtp configs, but SMTPConfigByAggregateID should return only one smtp config and that should be the one that has the active state (other smtp configs are inactive and only one can be active at any time): 

```
query, args, err := stmt.Where(sq.Eq{
		SMTPConfigColumnResourceOwner.identifier(): resourceOwner,
		SMTPConfigColumnInstanceID.identifier():    resourceOwner,
		SMTPConfigColumnState.identifier():         domain.SMTPConfigStateActive,
	}).ToSql()

```
If this solution acceptable? (sorry, as you know this is my second attempt for a serious backend change but it's obvious that I'm doing it wrong).

I was tempted to change the function name SMTPConfigByAggregateID with SMTPActiveConfig but I had some problems with MockGen to rename.

stebenz replied:
I would rename the function to clearly state the functionality. Additional question, if you only provide the SMTP config if active, is the handling of no config returned handled int he notification logic? In the API if you do a GET it's not important as you get back a NotFound, but the logic to notify changes as you also can have no active config.

doncicuto replied:
Thanks @stebenz. I've renamed SMTPConfigByAggregateID as SMTPConfigActive and run mockgen to remove SMTPConfigByAggregateID and include SMTPConfigActive in queries mocks.

For the second question, if no active config is found we follow the same behavior that Zitadel had when only one possible config was available, if no rows were returned in prepareSMTPConfigQuery the Errors.SMTPConfig.NotFound is thrown :

```
if err != nil {
	if errors.Is(err, sql.ErrNoRows) {
		return nil, zerrors.ThrowNotFound(err, "QUERY-fwofw", "Errors.SMTPConfig.NotFound")
	}
	return nil, zerrors.ThrowInternal(err, "QUERY-9k87F", "Errors.Internal")
}
`
And we see the same error in logs that in the past if no config is found:

```
ERRO[0151] statement execution failed                    caller="/home/doncicuto/Code/zitadel/internal/eventstore/handler/v2/handler.go:629" error="ID=QUERY-fwofw Message=Errors.SMTPConfig.NotFound Parent=(sql: no rows in result set)" projection=projections.notifications
`

If you want a different behavior let me know

stebenz replied:
That works for me

doncicuto commented on internal/query/smtp_test.go:
```diff
@@ -92,6 +99,9 @@ func Test_SMTPConfigsPrepares(t *testing.T) {
 						"host",
 						"user",
 						&crypto.CryptoValue{},
+						"2232323",
```
Hi @stebenz, I'll prepare them of course, thanks

stebenz replied:
Perfect, thanks

doncicuto commented on internal/query/smtp.go:
```diff
@@ -73,40 +84,71 @@ var (
+	ctx, span := tracing.NewSpan(ctx)
+	defer func() { span.EndWithError(err) }()
+
+	stmt, scan := prepareSMTPConfigQuery(ctx, q.client)
```
@stebenz I think I've answered this question in the previous conversation

stebenz replied:
Yes, will resolve this

doncicuto commented:
Hi @hifabienne FYI, I've updated the PR with @stebenz suggestions

peintnermax reviewed (dismissed):
Hi @doncicuto thx for the amazing work. I've found some minor things. Could you fix them and resolve the conflicts? That would be great! 😎

doncicuto commented:
Thanks @peintnermax I'll prepare the changes and update the PR as soon as posible 😉 Happy Easter! 🥚 🐰

peintnermax commented:
> Thanks @peintnermax I'll prepare the changes and update the PR as soon as posible 😉 Happy Easter! 🥚 🐰

Amazing.. thx a lot.. wish you happy easter too 😎👍

doncicuto commented on console/src/app/modules/policies/notification-sms-provider/password-dialog-sms-provider/password-dialog.component.ts:
Done!

doncicuto commented on console/src/app/pages/instance/instance-routing.module.ts:
```diff
@@ -32,6 +32,15 @@ const routes: Routes = [
+    canActivate: [AuthGuard, RoleGuard],
+    loadChildren: () => import('src/app/modules/smtp-provider/smtp-provider.module'),
+    data: {
+      roles: ['iam.idp.read'],
```
Done!

doncicuto commented on console/src/app/modules/policies/notification-sms-provider/notification-sms-provider.module.ts:
```diff
@@ -34,7 +34,6 @@ import { MatDialogModule } from '@angular/material/dialog';
     FormFieldModule,
     WarnDialogModule,
     MatSelectModule,
-    MatDialogModule,
```
Done!

doncicuto commented:
Hi @peintnermax I've updated the PR and I think I've solved conflicts and added requested changes and some more. As I created this PR some time ago sorry if there are some things that haven't aged well so just let me know any fix you need. Thanks!

peintnermax commented:
> Hi @peintnermax I've updated the PR and I think I've solved conflicts and added requested changes and some more. As I created this PR some time ago sorry if there are some things that haven't aged well so just let me know any fix you need. Thanks!

Amazing.. It looks good to me now.. 👍

fforootd commented:
Call me excited about this PR.

Thanks for all the hard work @doncicuto

stebenz commented on internal/command/instance_model.go:
```diff
@@ -81,6 +83,6 @@ func (wm *InstanceWriteModel) Query() *eventstore.SearchQueryBuilder {
 }
 
-func InstanceAggregateFromWriteModel(wm *eventstore.WriteModel) *eventstore.Aggregate {
+func InstanceAggregateFromWriteModel(ctx context.Context, wm *eventstore.WriteModel) *eventstore.Aggregate {
```
Please change this function back so that there is no context as parameter and have a look at for example ExecutionAggregateFromWriteModel to solve the linter issue.

doncicuto replied:
Ok, thanks @stebenz  for the hint, then do you mean I should use c.pushAppendAndReduce to fix the lint issues, right?

doncicuto replied:
Absolutely @stebenz

stebenz replied:
The linter issue for InstanceAggregateFromWriteModel came from the ctx which is no parameter right?
So if you change the content of the InstanceAggregateFromWriteModel function to be similar to ExecutionAggregateFromWriteModel without the ctx then a lot of changes in this PR should be reverted.

stebenz commented on internal/command/instance_smtp_config_model.go:
```diff
@@ -111,7 +103,13 @@ func (wm *InstanceSMTPConfigWriteModel) Reduce() error {
+func (wm *IAMSMTPConfigWriteModel) Query() *eventstore.SearchQueryBuilder {
+	// If ID equals ResourceOwner we're dealing with the old and unique smtp settings
+	// Let's set the empty ID for the query
+	if wm.ID == wm.ResourceOwner {
```
Instead of something like this, you could use '.EventData(map[string]interface{}{"id": wm.ID})' in the query directly. There are some examples for that also available in the code.

doncicuto replied:
Hi @stebenz, I'm not sure about this one. In my code for the old smtp settings I set the ID to the resourceowner value as we don't want the ID to be empty. 

Old events for the old smtp settings don't have the ID value set so if I want to filter those events I should look for events where ID is empty. If I use .EventData(map[string]interface{}{"id": wm.ID})' I wouldn't find the events as it would try to find events with ID=ResourceOwner...  This logic is implemented in internal/query/projection/smtp.go

```
	// Deal with old and unique SMTP settings (empty ID)
	id := e.ID
	description := e.Description
	state := domain.SMTPConfigStateInactive
	if e.ID == "" {
		id = e.Aggregate().ResourceOwner
		description = "generic"
		state = domain.SMTPConfigStateActive
```

What do you think I should do here @stebenz? Thanks in advance

stebenz replied:
You're right, ignore my comment here.

stebenz reviewed (commented):
@doncicuto I had another look, I think they are small chances that I missed but would be a good idea. Could you have a look if the comments are clear for you?

doncicuto commented on internal/api/grpc/admin/iam_settings_converter.go:
```diff
@@ -147,26 +148,29 @@ func AddSMTPToConfig(req *admin_pb.AddSMTPConfigRequest) *smtp.Config {
 
 func SMTPConfigToPb(smtp *query.SMTPConfig) *settings_pb.SMTPConfig {
 	mapped := &settings_pb.SMTPConfig{
+		Description:    smtp.Description,
```
Absolutely @stebenz

doncicuto commented:
Many thanks @stebenz and @peintnermax for reviewing this PR, I really hope this code works fine and looks good to you and that it will ease the configuration of different providers.

muhlemmer commented:
Thank you @doncicuto for this huge effort and seeing it through to the end! It's not easy when there are many reviewers involved :smile:

fforootd commented:
Thank you for this superb PR ❤️

Beatirce commented:
Thank you，How is Office365 configured?

Beatirce commented:
> Hi @Beatirce, unfortunately I couldn't add a template for Office/Microsoft365 as I have no account to test against it. I created a PR on friday to add an outlook.com template and a fix that should allow login auth for smtp communication with that provider, but I've no way to test against Office365. I'm on Discord so if you or anybody interested in this feature want to contact me and provide an account to test I'd love to try (I'm not sure if Oauth with SMTP will be needed).

Thanks,but I can't provide a test account,because I'm not postmaster,sorry.
